core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/transform/AbstractStreamingRequestMarshaller.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.transform;
+
+import static software.amazon.awssdk.http.Header.CHUNKED;
+import static software.amazon.awssdk.http.Header.CONTENT_LENGTH;
+import static software.amazon.awssdk.http.Header.TRANSFER_ENCODING;
+
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.runtime.transform.Marshaller;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+
+@SdkInternalApi
+public abstract class AbstractStreamingRequestMarshaller<T> implements Marshaller<T> {
+    protected final Marshaller<T> delegateMarshaller;
+    protected final boolean requiresLength;
+    protected final boolean transferEncoding;
+    protected final boolean useHttp2;
+
+    protected AbstractStreamingRequestMarshaller(Builder builder) {
+        this.delegateMarshaller = builder.delegateMarshaller;
+        this.requiresLength = builder.requiresLength;
+        this.transferEncoding = builder.transferEncoding;
+        this.useHttp2 = builder.useHttp2;
+    }
+
+    /**
+     * This method will run certain validations for content-length and add
+     * additional headers (like Transfer-Encoding) if necessary.
+     *
+     * If requiresLength and transferEncoding is not set to true and Content Length is missing,
+     * SDK is not required to calculate the Content-Length and delegate that behavior to the underlying http client.
+     *
+     * @param marshalled A mutable builder for {@link SdkHttpFullRequest} representing a HTTP request.
+     * @param contentLength Optional of content length
+     * @param requiresLength True if Content-Length header is required on the request
+     * @param transferEncoding True if "Transfer-Encoding: chunked" header should be set on request
+     * @param useHttp2 True if the operation uses http2
+     */
+    protected void addHeaders(SdkHttpFullRequest.Builder marshalled,
```

Comment:
nit, `final`

Reply:
I don't see the need for it. If the subclasses want to override the behavior, they should be able to do it. It doesn't matter in this case as these are all protected APIs and not used by customers.

Reply:
It's generally good practice to either document it if it's expected to be extended by subclass or mark it `final` if it's not intended to be overridden. I'd lean in towards to prohibiting it if it's not needed for now and we can loose it up in the future if needed.